lib/service/service.go

```diff
@@ -1405,26 +1424,82 @@ func (process *TeleportProcess) initDiagnosticService() error {
 	log := logrus.WithFields(logrus.Fields{
 		trace.Component: teleport.Component(teleport.ComponentDiagnostic, process.id),
 	})
-	var ready int64
+
+	var state int64
+
+	// The diagnostic.readyz service monitors for TeleportReadyEvent and updates
+	// the state accordingly.
 	process.RegisterFunc("diagnostic.readyz", func() error {
 		eventC := make(chan Event, 1)
 		process.WaitForEvent(process.ExitContext(), TeleportReadyEvent, eventC)
 		select {
 		case <-eventC:
-			atomic.StoreInt64(&ready, 1)
+			atomic.StoreInt64(&state, stateOK)
 			log.Infof("Detected that service started and joined the cluster successfully.")
 		case <-process.ExitContext().Done():
 			log.Debugf("Teleport is exiting, returning.")
 		}
 		return nil
 	})
 
+	// The degraded.monitor service monitors degraded and recovered events and
+	// updates the state accordingly.
+	process.RegisterFunc("degraded.monitor", func() error {
+		eventCh := make(chan Event, 1024)
+		process.WaitForEvent(process.ExitContext(), TeleportDegradedEvent, eventCh)
+		process.WaitForEvent(process.ExitContext(), TeleportOKEvent, eventCh)
+
+		var recoveryTime time.Time
+
+		for {
+			select {
+			case e := <-eventCh:
+				switch e.Name {
+				// If a degraded event was received, always change the state to degraded.
+				case TeleportDegradedEvent:
```

Comment:
this recovery logic is sound, but please factor it out to a separate function alongside with state and recovery time as it's not trivial any more